server/svix-server/src/v1/endpoints/stats.rs

```diff
@@ -0,0 +1,160 @@
+#[serde(rename_all = "camelCase")]
+struct AttemptStatisticsData {
+    // FIXME: skip if null?
+    success_count: Option<Vec<i64>>,
+    failure_count: Option<Vec<i64>>,
+}
+
+#[derive(Serialize, JsonSchema)]
+enum StatisticsPeriod {
+    #[allow(dead_code)]
+    OneDay,
+    FiveMinutes,
+}
+
+#[derive(Serialize, JsonSchema)]
+#[serde(rename_all = "camelCase")]
+struct AttemptStatisticsResponse {
+    data: AttemptStatisticsData,
+    // TODO: extract to a "flattened" struct. Common fields to any statistics responses.
+    period: StatisticsPeriod,
+    start_date: DateTime<Utc>,
+    end_date: DateTime<Utc>,
+}
+
+#[derive(Deserialize, JsonSchema)]
+#[serde(rename_all = "camelCase")]
+struct StatisticsParams {
+    start_date: Option<DateTime<Utc>>,
+    end_date: Option<DateTime<Utc>>,
+}
+
+/// Rows given by the app attempt stats query look like this.
+#[derive(Debug, FromQueryResult)]
+struct AppAttemptStatsResult {
+    /// 0 = success, 2 = failure
+    status: i16,
+    count: i64,
+    bucket: DateTime<Utc>,
+}
+
+/// Message attempt statistics for a given app.
+// FIXME: what's the operation id for this?
+#[aide_annotate(op_id = "v1.stats.fixme")]
+async fn get_app_attempt_stats(
+    State(AppState { ref db, .. }): State<AppState>,
+    // FIXME: should be like a uid or id. I'm sure there's an extractor for that or whatever.
+    Path(app_id): Path<String>,
+    Query(params): Query<StatisticsParams>,
+    // FIXME: any auth?
+) -> Result<Json<AttemptStatisticsResponse>, crate::error::Error> {
+    // XXX: Original looks like it provides a default of "now" and "6 hours ago" for end/start.
+    // I'm not sure what's meant to happen if you supply one or the other but not both.
+    // FIXME: There was a lot more date range validation in the original... I should check that out.
+    let end_date = params
+        .end_date
+        // XXX: the fallback in the original would round down to the nearest past 5 minutes, but I
+        // think this was just because the stats were being pre-calculated in buckets.
+        // Might not be needed here.
+        .unwrap_or_else(|| Utc::now());
+    let start_date = params
+        .start_date
+        .unwrap_or_else(|| end_date.sub(chrono::Duration::hours(6)));
+
+    // RE: date bounds:
+    // More simple to focus only on the ended_at since we know we only care about attempts
+    // that already have an outcome.
+    // If we need more sophisticated, we'd look for overlapping ranges of start/end vs created/ended.
+    // N.b. `BETWEEN` is inclusive on the upper bound, which is different than the original
+    // version of this.
+    let stmt = Statement::from_sql_and_values(
+        db.get_database_backend(),
+        r#"
+        SELECT
+            status,
+            count(1),
+            to_timestamp(
+                floor(EXTRACT(epoch FROM ended_at) / EXTRACT(epoch FROM interval '5 min'))
+                * EXTRACT(epoch FROM interval '5 min')
+            ) AS bucket
```

Comment:
This is pretty gnarly. Worth noting `date_bin` was added in postgres v14 and is the proper solution for this problem, but I don't know that we can force folks to upgrade their db (assuming they aren't already on v14).